lightning/src/util/ser_macros.rs

```diff
@@ -427,6 +454,139 @@ macro_rules! impl_writeable_tlv_based {
 	}
 }
 
+/// Defines a struct for a TLV stream and a similar struct using references for non-primitive types,
+/// implementing [`Readable`] for the former and [`Writeable`] for the latter. Useful as an
+/// intermediary format when reading or writing a type encoded as a TLV stream. Note that each field
+/// representing a TLV record has its type wrapped with an [`Option`].
+///
+/// [`Readable`]: crate::util::ser::Readable
+/// [`Writeable`]: crate::util::ser::Writeable
+macro_rules! tlv_stream {
+	($name:ident, $nameref:ident, {
+		$(($type:expr, $field:ident : $fieldty:ident$(<$gen:ident>)?)),* $(,)*
+	}) => {
+		#[derive(Debug)]
+		struct $name {
+			$(
+				$field: Option<$fieldty$(<$gen>)?>,
+			)*
+		}
+
+		pub(crate) struct $nameref<'a> {
+			$(
+				pub(crate) $field: Option<tlv_record_ref_type!($fieldty$(<$gen>)?)>,
+			)*
+		}
+
+		impl<'a> ::util::ser::Writeable for $nameref<'a> {
+			fn write<W: ::util::ser::Writer>(&self, writer: &mut W) -> Result<(), $crate::io::Error> {
+				encode_tlv_stream!(writer, {
+					$(($type, self.$field, (tlv_record, $fieldty$(<$gen>)?))),*
+				});
+				Ok(())
+			}
+		}
+
+		impl ::util::ser::Readable for $name {
+			fn read<R: $crate::io::Read>(reader: &mut R) -> Result<Self, ::ln::msgs::DecodeError> {
+				$(
+					init_tlv_field_var!($field, tlv_record);
+				)*
+				decode_tlv_stream!(reader, {
+					$(($type, $field, (tlv_record, $fieldty$(<$gen>)?))),*
+				});
+
+				Ok(Self {
+					$(
+						$field: init_tlv_based_struct_field!($field, tlv_record)
+					),*
+				})
+			}
+		}
+	}
+}
+
+macro_rules! tlv_record_ref_type {
+	(u8) => {
+		u8
+	};
+	(u16) => {
+		u16
+	};
+	(u32) => {
+		u32
+	};
+	(u64) => {
+		u64
+	};
+	(char) => {
+		char
+	};
+	(String) => {
+		&'a crate::prelude::String
+	};
+	(Vec<$type:ty>) => {
+		&'a crate::prelude::Vec<$type>
+	};
+	($type:ident$(<$gen:ident>)?) => {
+		&'a $type$(<$gen>)?
+	};
+}
+
+macro_rules! encoded_tlv_record_type {
+	(u8) => {
+		u8
+	};
+	(u16) => {
+		::util::ser::HighZeroBytesDroppedBigSize<u16>
+	};
+	(u32) => {
+		::util::ser::HighZeroBytesDroppedBigSize<u32>
+	};
+	(u64) => {
+		::util::ser::HighZeroBytesDroppedBigSize<u64>
+	};
+	(char) => {
+		char
+	};
+	(String) => {
+		::util::ser::WithoutLength<crate::prelude::String>
+	};
+	(Vec<$type:ty>) => {
+		::util::ser::WithoutLength<crate::prelude::Vec<$type>>
+	};
+	($type:ident$(<$gen:ident>)?) => {
+		$type$(<$gen>)?
+	};
+}
+
+macro_rules! encoded_tlv_record_ref_type {
+	(u8) => {
+		u8
+	};
+	(u16) => {
+		::util::ser::HighZeroBytesDroppedBigSize<u16>
```

Comment:
Are you suggesting we expand `encode_tlv` (the call site) such that there is a `tlv_record` matcher for each of these types? Could do that, but it would result in a lot of repetitive boilerplate. Also, wouldn't prevent doing thing surprising given the same matchers here would have to be used there.

Note that these three helper macros are only used in one place each. I think I could define them as nested macros at the respective calls sites. But then relative to each other they would be spread all about, and it's kinda nice having them all in one place.

Reply:
(Note for discussion that `u8` doesn't require `HighZeroBytesDroppedBigSize` since it is only one byte.)

Hmm... not sure if that's any better or all that straightforward. We'd still need special handling to distinguish between records that need a wrapper from those that don't (e.g., `u8`, `char`, and user-defined `struct`) when defining fields in each of the two TLV stream structs in the macro. This would require another (small) helper macro. Further, for the reference version, we'd still need to handle primitives differently, so `tlv_record_ref_type` would be needed still.

Additionally, for decoding (opposed to encoding), we aren't wrapping but rather specifying the type that must be used with `Readable` and then pulling out the inner value.

To you original question, the idea was for the `tlv_stream` macro to abstract away the encoding requirement of using minimal representations for integers and avoiding duplicating the length of strings and vectors. Seems making the caller explicitly state these puts the onus on them needlessly.

Reply:
Took a stab at this with a few caveats as noted below.

> Presumably that would also be at the callsite, so we'd still avoid the list of types in the macros, no? We'd just need two variants?

Yeah, though special handling is indeed needed for primitives (see below), though it shortens the list considerably.

> Not sure I understand why? We can serialize references to primitives, no?

Yes, but we can't return references to temporaries, like when converting from a `Duration` to `u64`. This is a more general problem whenever we represent a TLV record differently in the higher-level abstraction.

> So that wouldn't need to change? Maybe I misunderstood this point.

Yeah, I kept the `tlv_record` matchers for both actually. For encoding, simply wrapping as `$wrapper(self.$field)` actually doesn't work because `self.$field` is an `Option`.

> Right, it's the usual explicit-vs-magic tradeoff. Obviously I try to come down on the side of explicit as I'm no fan of magic. In serialization frameworks everything tends to be magic which always irritates me. Given the lightning spec keeps changing the serialization formats of things, resulting in a ton of different formats and wrappers, and our serialization test coverage (at least for internal stuff) is kinda poor, I worry about not seeing changes or understanding the exact behavior when reviewing future changes.

I don't have a strong opinion, to be honest, though my preference is more with the implicit approach. Seems like these are limited to variable length and primitive types. So I wouldn't expect this to grow with any spec changes.

Reply:
> For decoding, I'd kinda prefer we do something more generic. Instead of adding code to the general tlv macros just or the new `tlv_record`, we can add a new general-purpose "wrapper" write type. Instead of passing the `$fieldty` as-is to the TLV-read macro, we can first split the wrapper and the underlying type, and then just pass the wrapper itself through the ser_macro. As a bonus, we can eventually drop the `vec_type` stuff favor of the new wrapper (but that can be in a followup).

Not quite sure I follow what you are thinking here. Currently, `tlv_record` assumes the underlying field is an `Option`, but generally that is not the case and specifically is not the case for `vec_type`.

> Its pretty easy to avoid on the write side, eg: https://git.bitcoin.ninja/index.cgi?p=rust-lightning;a=commit;h=e6b3bc7f1cad9243c9cd9e281e8520ac57f83847

Pushed a variation of this that avoids needing `_tlv_stream_impl_writeable` and also removes `encoded_tlv_record_type` as a bonus.

Reply:
Putting `vec_type` aside, I can iterate on this to get rid of the `tlv_record` literal in favor of some variation of `option` that also takes a wrapper. This is pretty straightforward for `encode_tlv`. But for `decode_tlv` we already have a variation for specifying which `Readable*` trait along with optional args. So the question here is how do we want to formulate the matchers such that a `$fieldty:tt` can be passed through `decode_tlv_stream` to `decode_tlv` and handle either case.

Reply:
> Right, I don't think we want to pass the `$fieldty:tt` through, I think we'll want to split it like we do for the `_tlv_stream_impl_writeable` and then pass the wrapper through without the type itself. I had a patch to do this, but dropped it as I'd ended up rewriting more than necessary but it was pretty easy.

Ah, I mean how should we pass it through to disambiguate from the matcher taking a trait:

```rust
($reader: expr, $field: ident, (option: $trait: ident $(, $read_arg: expr)?)) => {{
```

There, `(option: $trait: ident $(, $read_arg: expr)?)` is a token tree needed by `decode_tlv_stream`. I get that we want to pull the wrapper from `$fieldty:tt`, as you stated, before giving it to `decode_tlv_stream`. My question is more about what the matcher should look like in `decode_tlv` to differentiate it from the above matcher.

Could just be something like:

```rust
($reader: expr, $field: ident, (option, $wrapper:ident)) => {{
```

But the only difference between that and the earlier one is using a comma instead of a colon and dropping the optional read arg, which isn't very intuitive. Maybe use a `wrapper` literal?

```rust
($reader: expr, $field: ident, (option, wrapper: $wrapper:ident)) => {{
```

Reply:
Ok, I removed the `tlv_record` literals but wasn't able to pull out the wrapper from `$fieldty:tt`. I guess it can be done like `_tlv_stream_impl_writeable`, but I was hoping to avoid the extra `()` around any types not needing special encoding. Drawback is anyone using this outside of `tlv_stream` would have to give both the encoding and the type. I suppose it is at least explicit and should fail to compile if given the wrong type.